consensus/obcpbft/events.go

```diff
@@ -0,0 +1,306 @@
+You may obtain a copy of the License at
+
+                 http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package obcpbft
+
+import "time"
+
+// event is an interface which defines an event which is identified by eventType
+type event interface {
+	eventType() eventType // The type of the event
+}
+
+// eventReceiver is a consumer of events, processEvent will be called serially
+// as events arrive
+type eventReceiver interface {
+	// processEvent delivers an event to the eventReceiver, if it returns non-nil, the return is the next processed event
+	processEvent(e event) event
+}
+
+// ------------------------------------------------------------
+//
+// Threaded object
+//
+// ------------------------------------------------------------
+
+// threaded holds an exit channel to allow threads to break from a select
+type threaded struct {
+	exit chan struct{}
+}
+
+// halt tells the threaded object's thread to exit
+func (t *threaded) halt() {
+	select {
+	case <-t.exit:
+		logger.Warning("Attempted to halt a threaded object twice")
+	default:
+		close(t.exit)
+	}
+}
+
+// ------------------------------------------------------------
+//
+// Event Manager
+//
+// ------------------------------------------------------------
+
+// eventManager provides a serialized interface for submitting events to
+// an eventReceiver on the other side of the queue
+type eventManager interface {
+	queue() chan<- event // Get a write-only reference to the queue, to submit events
+	start()              // Starts the eventManager thread TODO, these thread management things should probably go away
+	halt()               // Stops the eventManager thread
+}
+
+// eventManagerImpl is an implementation of eventManger
+type eventManagerImpl struct {
+	threaded
+	receiver eventReceiver
+	events   chan event
+}
+
+// newEventManager creates an instance of eventManagerImpl
+func newEventManagerImpl(er eventReceiver) eventManager {
+	return &eventManagerImpl{
+		receiver: er,
+		events:   make(chan event),
+		threaded: threaded{make(chan struct{})},
+	}
+}
+
+// start creates the go routine necessary to deliver events
+func (em *eventManagerImpl) start() {
+	go em.eventLoop()
+}
+
+// queue returns a write only reference to the event queue
+func (em *eventManagerImpl) queue() chan<- event {
+	return em.events
+}
+
+// eventLoop is where the event thread loops, delivering events
+func (em *eventManagerImpl) eventLoop() {
+	for {
+		select {
+		case next := <-em.events:
+			// If an event returns something non-nil, then process it as a new event
+			for ; next != nil; next = em.receiver.processEvent(next) {
```

Comment:
I have no very strong feelings here, the code as written is shorter, and follow normal for loop construction, but, if this is bothersome, it can be changed.